Fix copyrights.csv quoting and extra fields

pehjota opened the pull request:
The first `update_copyrights` run writes output without quoting fields that contain commas, so the second run parses them as multiple fields and adds a "missing" extra field to each of all the other rows.  The first commit fixes this by using a `csv.writer()` object instead of `str.join()` to format rows for output.  It also undoes the removal of commas from fields that was necessary in commit 676c1fa2b98.

The second commit adds a check for extra or missing fields, which would have caught the aforementioned issue but now caught another one: three rows added in commit c6313453143c had duplicated MD5 fields.

stevecotton commented on update_copyrights:
```diff
@@ -120,30 +124,32 @@ for root, _, files in os.walk(options.repo):
-if len(added) > 0 or len(changed) > 0 or len(incomplete) > 0 or len(update) > 0:
+if len(malformed) > 0 or len(added) > 0 or len(changed) > 0 or len(incomplete) > 0 or len(update) > 0:
     any_check_failed = True
+    print("There are "+str(len(malformed))+" rows with too many or too few fields, possibly due to an unquoted comma")
```
My gut feeling is that this should go between "incomplete" and "update", rather than at the top.

pehjota replied:
My gut feeling was that malformed rows that can't be parsed correctly (and therefore might also be changed, incomplete, or outdated, because the script might be misreading one field as another) are more severe than those with field values that are parsable but simply in need of updates.

stevecotton reviewed (approved):
I'd be happy if this PR merged as-is.

I feel there's room to improve the error messages and error detection, for example to say whether it's missing a field, and only suggesting that there's an unquoted comma when there are too many fields, but that's probably for a later PR. This one fixes the handling of proper CSV, and a separate PR about better error messages can enjoy time for bikeshedding.

pehjota commented on update_copyrights:
```diff
@@ -66,6 +66,8 @@ with contextlib.suppress(FileNotFoundError):
+malformed = []
 added = []
 changed = []
 # Already mentioned in the CSV file, but lacking something in either the license or author fields
```
I had the same thought but wanted to keep the variables declared in order of use.  Adding comments to `added` and `changed` is a good idea.  Another option is to just put the comment for `malformed` at the end of the declaration line, since it's short anyway, but that breaks the existing style.

pehjota commented:
> I feel there's room to improve the error messages and error detection, for example to say whether it's missing a field, and only suggesting that there's an unquoted comma when there are too many fields, but that's probably for a later PR. This one fixes the handling of proper CSV, and a separate PR about better error messages can enjoy time for bikeshedding.

That's a good idea and would be easy to do by splitting `malformed` into two lists, e.g. `extra_fields` (possible unquoted comma in a field) and `missing_fields`.

pehjota commented:
I split `malformed` into two lists for better error messages and added comments to `added` and `changed` declarations, but I didn't move the too few/many fields error messages.  If you feel strongly that having correct field values is more important than first knowing which if any field values are incorrect, I can change that.

stevecotton commented:
I'll leave it a while for anyone else to comment, but am planning to merge it tomorrow unless someone says otherwise.

pehjota commented:
Sounds good, thanks.
